Add RowNumber spill partition bits

duanmeng opened the pull request:
Add `rowNumberPartitionBits` in SpillConfig for RowNumber,
and consolidate the `spillLevel` and `exceedMaxSpillLevel`
methods.

xiaoxmeng commented on velox/common/base/SpillConfig.h:
```diff
@@ -69,10 +70,17 @@ struct SpillConfig {
   bool exceedJoinSpillLevelLimit(uint8_t startBitOffset) const;
 
+  /// TODO: Remove above method
+  bool exceedSpillLevelLimit(uint8_t startBitOffset, uint8_t partitionBits)
```
Can you replace in this PR? It doesn't break anything? Thanks!

duanmeng replied:
Fixed

xiaoxmeng commented on velox/common/base/SpillConfig.h:
```diff
@@ -69,10 +70,17 @@ struct SpillConfig {
   int32_t joinSpillLevel(uint8_t startBitOffset) const;
 
+  /// TODO: Remove above method
+  int32_t spillLevel(uint8_t startBitOffset, uint8_t partitionBits) const;
```
ditto

xiaoxmeng commented on velox/common/base/tests/SpillConfigTest.cpp:
```diff
@@ -37,6 +37,7 @@ TEST(SpillConfig, spillLevel) {
       0,
       kInitialBitOffset,
       kNumPartitionsBits,
```
Can you add tests for the new APIs? Thanks!

duanmeng replied:
Got it, added.

xiaoxmeng reviewed (approved):
@duanmeng LGTM. Thanks!

duanmeng commented:
Follow-up: Add spill related session property for native engine refering https://github.com/prestodb/presto/pull/21836

xiaoxmeng commented on velox/common/base/SpillConfig.h:
```diff
@@ -57,21 +57,24 @@ struct SpillConfig {
       const std::string& _fileCreateConfig = {});
 
-  /// Returns the hash join spilling level with given 'startBitOffset'.
+  /// Returns the hash join spilling level with given 'startBitOffset' and
```
Update the comment?

xiaoxmeng commented on velox/common/base/SpillConfig.h:
```diff
@@ -113,13 +116,17 @@ struct SpillConfig {
   int32_t spillableReservationGrowthPct;
 
-  /// Used to calculate spill partition number.
+  /// The started partition bit offset of the root level (level 0) partitions.
```
```
s/root/top (the first level)/
```

xiaoxmeng commented on velox/common/base/SpillConfig.h:
```diff
@@ -113,13 +116,17 @@ struct SpillConfig {
   /// 'startPartitionBit'.
   uint8_t joinPartitionBits;
 
+  /// Used to calculate the spill hash partition number for RowNumber with
```
```
/// Used to calculate the spill partition number of the hash table in RowNumber with 'startPartitionBit'.
```

xiaoxmeng commented on velox/common/base/SpillConfig.cpp:
```diff
@@ -59,8 +61,9 @@ SpillConfig::SpillConfig(
-  const auto numPartitionBits = joinPartitionBits;
+int32_t SpillConfig::spillLevel(uint8_t startBitOffset, uint8_t partitionBits)
+    const {
+  const auto numPartitionBits = partitionBits;
```
Why we need numPartitionBits?

xiaoxmeng commented on velox/common/base/SpillConfig.h:
```diff
@@ -57,21 +57,24 @@ struct SpillConfig {
   /// NOTE: we advance (or right shift) the partition bit offset when goes to
   /// the next level of recursive spilling.
-  int32_t joinSpillLevel(uint8_t startBitOffset) const;
+  int32_t spillLevel(uint8_t startBitOffset, uint8_t partitionBits) const;
```
this shall be a static method?

xiaoxmeng reviewed (commented):
@duanmeng LGTM % minors. Thanks!

tanjialiang commented on velox/common/base/SpillConfig.h:
```diff
@@ -113,13 +116,17 @@ struct SpillConfig {
   int32_t spillableReservationGrowthPct;
 
-  /// Used to calculate spill partition number.
+  /// The started partition bit offset of the top (the first level) partitions.
```
started -> start

duanmeng replied:
Fixed

tanjialiang commented on velox/core/QueryConfig.h:
```diff
@@ -589,6 +592,18 @@ class QueryConfig {
         kMaxBits, get<uint8_t>(kJoinSpillPartitionBits, kDefaultBits));
   }
 
+  /// Returns the number of bits used to calculate the spilling partition
+  /// number for RowNumber. The number of spilling partitions will be power of
+  /// two.
+  ///
+  /// NOTE: as for now, we only support up to 8-way spill partitioning.
```
spilling -> spill

duanmeng replied:
Fixed, and fix the comments of `joinSpillPartitionBits()` too.

mbasmanova commented on velox/common/base/SpillConfig.h:
```diff
@@ -113,13 +116,17 @@ struct SpillConfig {
 
+  /// Used to calculate the spill partition number of the hash table in
+  /// RowNumber with 'startPartitionBit'.
+  uint8_t rowNumberPartitionBits;
```
Curious, why do we want separate variable for join and rowNumber?

duanmeng replied:
This would enable us to fine-tune the number of spill partitions per operator.

mbasmanova replied:
Why would we want to do that?

duanmeng replied:
At first, I want to differentiate the `joinPartitionBits` and `rowNumberPartitionBits` when calling the function `exceedSpillLevelLimit`, so I added the config `rowNumberPartitionBits`.
e.g.
- HashBuild
  - spillConfig.exceedSpillLevelLimit(startBit, **spillConfig.joinPartitionBits**)
- RowNumber
  - spillConfig.exceedSpillLevelLimit(startBit, **spillConfig.rowNumberPartitionBits**)

duanmeng replied:
> Yes, I don't see any reason for setting different values of numPartitionBits for different operators and I'm a bit concerned about having inconsistent values.

Got it, I will unify them as `numPartitionBits`.

mbasmanova commented on velox/core/QueryConfig.h:
```diff
@@ -273,6 +273,9 @@ class QueryConfig {
       "join_spiller_partition_bits";
 
+  static constexpr const char* kRowNumberSpillPartitionBits =
+      "row_number_spiller_partition_bits";
```
Please, update documentation.

duanmeng replied:
Got it, fixed.

mbasmanova reviewed (commented):
@duanmeng When we spill for the first time, the data is written to disk and read back once. When we spill again (next level), the data is written to disk and read back again. It seems like the data is written to disk and read as many times as there are spill levels. Is this accurate?

duanmeng commented:
> @duanmeng When we spill for the first time, the data is written to disk and read back once. When we spill again (next level), the data is written to disk and read back again. It seems like the data is written to disk and read as many times as there are spill levels. Is this accurate?

@mbasmanova Hi Masha, thanks for your review. The first spill writes all the data to disk. After this, the RowNumber restores the data from one partition per time, clearing each partition's data once it has been processed (#8413). In the event of a recursive spill, only the currently restored partition data in memory is spilled to disk (next level), and only one partition from the next level is read back into memory.

> It seems like the data is written to disk and read as many times as there are spill levels. 

This scenario only occurs in the absolute worst case, where a spill (recursive) is triggered during the processing of each restored partition.

duanmeng commented:
@mbasmanova Hi Masha, I've removed the rowNumberPartitionBits and renamed the joinPartitionBits to numPartitionBits according to your comments. Could you please help take another look? Thanks.
